### ISSUE TITLE
Native validation doesn't dismiss errors as the user corrects them

### ISSUE THREAD

**Issue by sbking** [adobe/react-spectrum#6983]
### Provide a general summary of the issue here

React ARIA's "native" validation doesn't work like HTML5 native validation works in most browsers, which results in a significantly worse user experience.

### 🤔 Expected Behavior?

In most modern browsers, when an HTML5 validation error is already showing, the validation error will update in real time until the user has fixed it, and then the error will immediately go away without the user blurring the field.

### 😯 Current Behavior

When `validationBehavior` is set to `"native"` and a validation error is showing, the error will never go away until the user blurs the field. The error message also won't update if the native validation issue changes until the user blurs the field.

### 💁 Possible Solution

I think the `useFormValidation` component should have two modes:

- A "lazy" mode which is initially active - in this mode it would only call `state.commitValidation()` in the `change` event handler
- An "eager" mode which becomes active while a native validation error is showing - in this mode if would call `state.commitValidation()` in an `input` event handler

### 🖥️ Steps to Reproduce

https://codesandbox.io/s/quirky-meitner-jl79fh?file=/src/App.js

### Version

latest

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari

### What operating system are you using?

macOS Sonoma 14.3

**Comment by LFDanLu** [comment]
The behavior of validating on blur for "native" validation is intentional since we thought it might be distracting to the user to display errors as they were entering their updated input value. Does implementing realtime validation like as shown [here](https://react-spectrum.adobe.com/react-aria/forms.html#realtime-validation) suffice instead?

**Comment by musjj** [comment]
Yep, I tested the native form in both Firefox and Chrome. After an error, the field switches to real-time validation and updates the error as the user types. Would love to see the same behavior with `react-aria-component`'s `Form`s.

**Comment by devongovett** [comment]
The goal isn't to match what browsers do by default - for that you don't even need React Aria at all, you can just use native inputs. Our goal is to improve the UX. Showing errors while the user is typing and has not completed their input yet is very distracting and can be misleading. An error might appear and disappear multiple times, for example while entering an email address, which could be very confusing for users. That's why by default we show errors only on "complete" input, which we determine from the blur event.

If you want different behavior, there is support for completely controlling the live error message using props, or using a library like react-hook-form.

**Comment by musjj** [comment]
> The goal isn't to match what browsers do by default - for that you don't even need React Aria at all, you can just use native inputs

I'd agree that strictly matching the browser's behavior just for the sake of it isn't good. But I think the standard browser behavior in this case is just really nice.

To clarify, errors are *only* re-validated in real-time *after* a submit attempt that triggers a validation error (you can check the code sandbox above). When typing for the first time, no error messages will be displayed until the user submits the form.

> An error might appear and disappear multiple times

With the native form behavior, the field stops re-validating in real time after valid input is typed in. So once the error disappears, it will not show up again until the user attempts to submit (or blur). IMO this behavior is super reasonable and very user-friendly.

**Comment by devongovett** [comment]
Looks like all browsers implement this differently. [Here's another example](https://codepen.io/devongovett/pen/vYqbRNW).

* In Safari, errors clear immediately as soon as the user edits the value, even if the new value is still invalid. Revalidation only occurs when the user resubmits. This is a bit misleading since the value is still invalid even though it looks like it isn't. For example, if the error is "Please enter at least 8 characters" and there are two characters in the field, entering a third character hides the error even though it is still invalid.
* In Chrome, the error message updates as you type after an error. If you type just "devon" into the email field and hit submit, you'll get an error: "Please include an @ in the email address". Then typing the "@" immediately changes the error to "Please enter a part following '@'". This is annoying because I'm not done typing yet.
* In Firefox, the error doesn't disappear until the field becomes valid, but the error message does not update as you type. For example, typing one character into the min length field errors with "Please use at least 8 characters (you are currently using 1 characters)". Adding another character doesn't update the message.

None of these are ideal in my opinion. I think Safari's behavior is the least annoying, but still misleading. Firefox's behavior would be better if the error message was less specific (e.g. not including the current character count). We can't control that though (the default error messages come from the browser).

I think I'd want to continue revalidating on blur rather than only on re-submit, and not updating the error message live. Not sure whether it's better to hide the previous error immediately like Safari and potentially re-display it on blur, or just wait until the user is done editing to update it like we do today.

**Comment by devongovett** [comment]
I agree that showing outdated error messages is not ideal. It's just a question of when exactly to hide/update the error. As I showed above, there isn't a clear standard across browsers for this. Each option has pros and cons.

I could see potentially implementing Safari's behavior of hiding the error message as soon as the user edits the value, but as I said before it's still not ideal if the value is actually still invalid and the error re-appears on blur. It could also cause additional layout shifts while typing.

If the error messages were less specific (e.g. "Please enter an email address" instead of "Please include an @ in the email address") then we could potentially wait until the input is valid to hide the message since it wouldn't change as you type, but we can't guarantee this since the messages depend on the browser.

Updating the error message live has a number of downsides:

* As mentioned in the articles you linked, it's annoying to see error messages about incomplete input.
* For screen reader users, hearing a potentially irrelevant error message after typing each character rather than the character you entered could be very confusing.
* If the error message changes length, it could cause a lot of layout shifts when not shown as a floating overlay.

Another thing to consider is what to do with errors that come from server side validation. For example, say you enter an email address in a signup form and submit, and then the server says there is already an account with that email. There's not really a way to know when to hide that error except by re-submitting the form. Perhaps that's another reason to go with Safari's behavior of hiding as soon as the value changes. We currently clear server errors on blur. Would be nice if errors from client side and server side validation behaved somewhat consistently.

**Comment by nathanforce** [comment]
Following along here..is there a current suggested solution for showing validation errors in real time only once a field has been committed with an invalid value? Switching all fields/error messages to controlled does not seem like a tenable solution.

It seems that switching to `validationBehavior="aria"` allows for the desired ux except that it does not block form submissions when there is an invalid field.  

Where does this leave us? Are we stuck choosing between one of these (imo) poor user experiences or re-implementing the whole of form validation ourselves?

My ideal requirements:
* I can use uncontrolled form components, optionally defining constraints via `validate()`.
* Support for server validation
* Submits are prevented when the form is invalid
